rholang/src/test/scala/coop/rchain/rholang/interpreter/MatchTest.scala

```diff
@@ -273,4 +274,138 @@ class VarMatcherSpec extends FlatSpec with Matchers {
+      Send(Quote(GInt(1)), Seq(GInt(2)), persistent = false),
+      Send(Quote(GInt(2)), Seq(GInt(3)), persistent = false),
+      Send(Quote(GInt(3)), Seq(GInt(4)), persistent = false)
+    )
+    // ~Nil
+    val pattern: Connective = Connective(ConnNotBody(Par()))
+
+    val expectedResult = Some(Map.empty[Int, Par])
+    val result         = spatialMatch(target, pattern).runS(emptyMap)
+
+    result should be(expectedResult)
+
+    val patternPar: Par = pattern
+    val parResult       = spatialMatch(target, patternPar).runS(emptyMap)
+    parResult should be(expectedResult)
+
+    // ~Nil | ~Nil
+    val doublePatternPar = patternPar.addConnectives(pattern)
+    val doubleResult     = spatialMatch(target, doublePatternPar).runS(emptyMap)
+    doubleResult should be(expectedResult)
+
+    // ~Nil | ~Nil | ~Nil
+    val triplePatternPar = doublePatternPar.addConnectives(pattern)
+    val tripleResult     = spatialMatch(target, triplePatternPar).runS(emptyMap)
+    tripleResult should be(expectedResult)
+
+    // ~Nil | ~Nil | ~Nil | ~Nil
+    // Fails because there is no way to split 3 sends into 4 non nil terms.
+    val quadruplePatternPar = triplePatternPar.addConnectives(pattern)
+    val quadrupleResult     = spatialMatch(target, quadruplePatternPar).runS(emptyMap)
+    quadrupleResult should be(None)
+  }
+
+  "Matching a complicated connective" should "work" in {
+    // @1!(6) | @2!(7) | @3!(8)
+    val target: Par = Par().addSends(
+      Send(Quote(GInt(1)), Seq(GInt(6)), persistent = false),
+      Send(Quote(GInt(2)), Seq(GInt(7)), persistent = false),
+      Send(Quote(GInt(3)), Seq(GInt(8)), persistent = false)
+    )
+
+    val failTarget: Par = Par().addSends(
+      Send(Quote(GInt(1)), Seq(GInt(6)), persistent = false),
+      Send(Quote(GInt(2)), Seq(GInt(9)), persistent = false),
+      Send(Quote(GInt(3)), Seq(GInt(8)), persistent = false)
+    )
+
+    // ~Nil
+    val nonNullConn: Connective = Connective(ConnNotBody(Par()))
+    val nonNull: Par            = nonNullConn
+
+    // ~{ ~Nil | ~Nil }
+    val singleFactor: Connective = Connective(
+      ConnNotBody(
+        Par()
+          .addConnectives(nonNullConn, nonNullConn)
+          .withConnectiveUsed(true)))
+    // ~{ ~Nil | ~Nil } & ~Nil
+    val prime: Connective = Connective(
+      ConnAndBody(
+        ConnectiveBody(Seq(nonNull, Par().addConnectives(singleFactor).withConnectiveUsed(true)))))
+    // x /\ y!(7)
+    val capture: Connective =
+      Connective(
+        ConnAndBody(ConnectiveBody(
+          Seq(EVar(FreeVar(0)), Send(ChanVar(FreeVar(1)), Seq(GInt(7))).withConnectiveUsed(true)))))
+    // x!(7) \/ x!(8)
+    val alternative: Connective = Connective(
+      ConnOrBody(
+        ConnectiveBody(Seq(Send(ChanVar(FreeVar(0)), Seq(GInt(7))).withConnectiveUsed(true),
+                           Send(ChanVar(FreeVar(0)), Seq(GInt(8))).withConnectiveUsed(true)))))
+    // ~{ ~Nil | ~Nil } & ~Nil | x /\ y!(7) | x!(7) \/ x!(8)
```

Comment:
Is x used freely in this formula?  If so, how does x bind in the AND when it appears in the OR?  Are free variables just treated like wildcards in an OR?

Reply:
They aren't treated exactly like wildcards. They only match if they **could** capture, but they don't capture.